src/libraries/System.Private.CoreLib/src/System/Half.cs

```diff
@@ -0,0 +1,707 @@
+        /// </summary>
+        /// <param name="s">The input to be parsed.</param>
+        /// <param name="style">The <see cref="NumberStyles"/> used to parse the input.</param>
+        /// <param name="provider">A format provider. </param>
+        /// <param name="result">The equivalent <see cref="Half"/> value representing the input string if the parse was successful. If the input exceeds Half's range, a <see cref="Half.PositiveInfinity"/> or <see cref="Half.NegativeInfinity"/> is returned. If the parse was unsuccessful, a default <see cref="Half"/> value is returned.</param>
+        /// <returns><see langword="true" /> if the parse was successful, <see langword="false" /> otherwise.</returns>
+        public static bool TryParse(string s, NumberStyles style, IFormatProvider? provider, out Half result)
+        {
+            return TryParse(s.AsSpan(), style, provider, out result);
+        }
+
+        /// <summary>
+        /// Tries to parse a <see cref="Half"/> from a <see cref="ReadOnlySpan{Char}"/> with the given <see cref="NumberStyles"/> and <see cref="IFormatProvider"/>.
+        /// </summary>
+        /// <param name="s">The input to be parsed.</param>
+        /// <param name="style">The <see cref="NumberStyles"/> used to parse the input.</param>
+        /// <param name="provider">A format provider. </param>
+        /// <param name="result">The equivalent <see cref="Half"/> value representing the input string if the parse was successful. If the input exceeds Half's range, a <see cref="Half.PositiveInfinity"/> or <see cref="Half.NegativeInfinity"/> is returned. If the parse was unsuccessful, a default <see cref="Half"/> value is returned.</param>
+        /// <returns><see langword="true" /> if the parse was successful, <see langword="false" /> otherwise.</returns>
+        public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatProvider? provider, out Half result)
+        {
+            bool ret = false;
+            if (float.TryParse(s, style, provider, out float floatResult))
+            {
+                result = (Half)floatResult;
+                ret = true;
+            }
+            else
+            {
+                result = default(Half);
+            }
+            return ret;
+        }
+
+        private static bool AreZero(Half left, Half right)
+        {
+            // IEEE defines that positive and negative zero are equal, this gives us a quick equality check
+            // for two values by or'ing the private bits together and stripping the sign. They are both zero,
+            // and therefore equivalent, if the resulting value is still zero.
+            return (ushort)((left.m_value | right.m_value) & ~SignMask) == 0;
+        }
+
+        private static bool IsNaNOrZero(Half value)
+        {
+            return ((value.m_value - 1) & ~SignMask) >= PositiveInfinityBits;
+        }
+
+        private static ushort StripSign(Half value)
```

Comment:
You may be losing performance by using `ushort` for temporary results, using uint for scratch results may be better.

Have you looked at the JITed code and whether it is reasonably good?

Reply:
Yup, you're right. Returning a `uint` is slightly more efficient.